DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/autorun/src/autorun.cpp

```diff
@@ -0,0 +1,170 @@
+#include <CL/sycl.hpp>
+#include <sycl/ext/intel/fpga_extensions.hpp>
+
+// dpc_common.hpp can be found in the dev-utilities include folder.
+// e.g., $ONEAPI_ROOT/dev-utilities//include/dpc_common.hpp
+#include "dpc_common.hpp"
+
+#include "autorun.hpp"
+
+#include <algorithm>
+#include <iostream>
+#include <vector>
+
+using namespace sycl;
+
+// choose the device selector based on emulation or actual hardware
+// we make this a global variable so it can be used by the autorun kernels
+#if defined(FPGA_EMULATOR)
+  ext::intel::fpga_emulator_selector ds;
+#else
+  ext::intel::fpga_selector ds;
+#endif
+
+// declare the kernel names globally to reduce name mangling
+class ARProducerId;
+class ARKernelID;
+class ARConsumerId;
+class ARForeverProducerId;
+class ARForeverKernelID;
+class ARForeverConsumerId;
+
+// declare the pipe names globally to reduce name mangling
+class ARProducePipeID;
+class ARConsumePipeID;
+class ARForeverProducePipeID;
+class ARForeverConsumePipeID;
+
+// pipes
+using ARProducePipe = ext::intel::pipe<ARProducePipeID, int>;
+using ARConsumePipe = ext::intel::pipe<ARConsumePipeID, int>;
+using ARForeverProducePipe = ext::intel::pipe<ARForeverProducePipeID, int>;   
+using ARForeverConsumePipe = ext::intel::pipe<ARForeverConsumePipeID, int>;
+
+////////////////////////////////////////////////////////////////////////////////
+// Autorun user kernel and global variable
+struct MyAutorun {
+  void operator()() const {
+    // notice that in this version, we explicitly add the while(1)-loop
```

Comment:
This implies that an autorun kernel must run forever, either with an explicit or an implicit while(1).  Are there use cases for Autorun where the kernel might do a limited amout of stuff then finish?  I can't think of anything useful off the top of my head . . .

Reply:
Joe wanted these because he likes the explicit while-loop. And it at least gives the user an option to not have the while(1)